### PR TITLE
Experimental packaging

### PR DESCRIPTION
**Pull request by larioj**
This PR merges the new packaging functionality. It provides a new subcommand called experimental, along with three new packaging commands `dcos experimental package add`, `dcos experimental package build`, and `dcos experimental service start`. It also provides tests, although a large portion of them are commented out since some of the required functionality is not enabled in cosmos.

### CODE REVIEW

File: cli/dcoscli/experimental/main.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+            raise DCOSException(
+                'Output file [{}] already exists'.format(
+                    dcos_package_path))
+
+        # create a new file to contain the package
+        temp_file.file.seek(0)
+        with util.open_file(dcos_package_path, 'w+b') as dcos_package:
+            shutil.copyfileobj(temp_file.file, dcos_package)
+
+    print('Created DCOS Universe package: ', file=sys.stderr, end='')
```
Comment (tamarrow) [inline cli/dcoscli/experimental/main.py]: use emitter.publish or logger.debug

File: cli/dcoscli/experimental/main.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
+from __future__ import print_function
```
Comment (tamarrow) [inline cli/dcoscli/experimental/main.py]: we should be using `emitting`

File: cli/tests/data/package_build/helloworld/config.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (tamarrow) [inline cli/tests/data/package_build/helloworld/config.json]: extra fike: `cli/tests/data/package_build/helloworld/.DS_Store`

File: cli/tests/integrations/test_experimental.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+    assert_command(command, stdout=out)
+
+
+def test_version():
+    command = command_base + ['--version']
+    out = b'dcos-experimental version SNAPSHOT\n'
+    assert_command(command, stdout=out)
+
+
+def test_package_build_with_only_resources_with_only_resources_reference():
```
Comment (tamarrow) [inline cli/tests/integrations/test_experimental.py]: `with_only_resources` is twice

File: cli/tests/integrations/test_experimental.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+    _successful_package_build_test(
+        os.path.join(
+            build_data_dir,
+            "package_config_reference_expected.json"),
+        expected_package_path=os.path.join(
+            build_data_dir,
+            "package_config_reference_expected.json"))
+
+
+def test_package_build_with_only_config_with_only_config_reference():
```
Comment (tamarrow) [inline cli/tests/integrations/test_experimental.py]: `with_only_config` is twice

File: cli/tests/integrations/test_experimental.py [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+    _successful_package_build_test(
+        os.path.join(
+            build_data_dir,
+            "package_config_reference.json"),
+        expected_package_path=os.path.join(
+            build_data_dir,
+            "package_config_reference_expected.json"))
+
+
+def test_package_build_with_only_marathon_with_only_marathon_reference():
```
Comment (tamarrow) [inline cli/tests/integrations/test_experimental.py]: `with_only_marathon`

File: cli/tests/integrations/test_experimental.py [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+    assert parts is not None, package_path
+    return parts.group(1), parts.group(2), parts.group(3)
+
+
+def _get_md5_hash(path):
+    with open(path, 'rb') as f:
+        return util.md5_hash_file(f)
+
+
+def _get_json(path):
```
Comment (tamarrow) [inline cli/tests/integrations/test_experimental.py]: see `integrations/common::file_json_ast`

File: cli/tests/integrations/test_experimental.py [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+    with open(path) as f:
+        file_contents = f.read()
+    return json.loads(file_contents)
+
+
+def _get_json_in_zip(path, inner_file):
```
Comment (tamarrow) [inline cli/tests/integrations/test_experimental.py]: zip_contents_as_json

**Comment by tamarrow** [comment]
Have you run the skipped tests locally? We should have them run if an env var is set.

**Comment by larioj** [comment]
@tamarrow Yes.

File: cli/tests/integrations/test_experimental.py [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+            '{}-{} is already in progress\n'.format(
+                name, version).encode())
+        done_adding = code == 0
+        assert (done_adding or
+                (code == 1 and
+                 err == change_in_progress_message))
+        assert retry_number != max_retries, \
+            'Waiting for package add to complete took too long'
+        retry_number += 1
+        time.sleep(5)
```
Comment (tamarrow) [inline cli/tests/integrations/test_experimental.py]: how long is this expected to take? waiting a minute for a test is a long time

Comment (larioj) [inline cli/tests/integrations/test_experimental.py]: Running on my machine, this seemed to be about double the amount it usually took to complete.

Comment (tamarrow) [inline cli/tests/integrations/test_experimental.py]: I can't parse that. How long does it take to complete?

Comment (larioj) [inline cli/tests/integrations/test_experimental.py]: It's expected to take about 15 seconds. It usually gets it at the third request.

Comment (tamarrow) [inline cli/tests/integrations/test_experimental.py]: thats a long time...
can we use a simpler example to speed this up?

File: cli/dcoscli/data/help/experimental.txt [hunk 10]
```diff
@@ -0,0 +1,51 @@
+Description:
+    Experimental commands. These commands are under development and are subject to change.
```
Comment (orsenthil) [inline cli/dcoscli/data/help/experimental.txt]: Is there a JIRA ticket to track and read about the motivation.
* Experimentation via beta-build is a common/ well-known approach in projects that are open source.
* Adding a sub-command to public releases will make people rely on it and will be hard to go back on experiments. ( I hope that was vetted out).

File: cli/tests/integrations/test_experimental.py (hunk 9 above)
Comment (larioj) [inline cli/tests/integrations/test_experimental.py]: No

File: cli/dcoscli/data/help/experimental.txt (hunk 10 above)
Comment (tamarrow) [inline cli/dcoscli/data/help/experimental.txt]: cc @pyronicide @jsancio

**Comment by larioj** [comment]
Closing this to avoid permission issues. See for #858 for the continuation.

**Pull request closed without merging**